### PR TITLE
Add sparse categorical focal crossentropy

### LINKED ISSUE #23442: Feature request: Add SparseCategoricalFocalCrossentropy
## Summary

Add sparse-label support for categorical focal crossentropy through:

- `keras.losses.sparse_categorical_focal_crossentropy`
- `keras.losses.SparseCategoricalFocalCrossentropy`

Keras currently provides `CategoricalFocalCrossentropy`, which expects one-hot labels, and `SparseCategoricalCrossentropy`, which accepts integer labels. There is no corresponding focal loss for sparse categorical labels.

## Motivation

Sparse labels are common in semantic segmentation, token classification, and multiclass problems with many classes. Converting these labels to one-hot tensors solely to use focal loss increases memory usage and adds unnecessary preprocessing.

A native sparse implementation would provide focal loss while preserving integer labels and Keras' backend-agnostic behavior.

## Proposed API

```python
keras.losses.SparseCategoricalFocalCrossentropy(
    alpha=0.25,
    gamma=2.0,
    from_logits=False,
    ignore_class=None,
    axis=-1,
    reduction="sum_over_batch_size",
    name="sparse_categorical_focal_crossentropy",
    dtype=None,
)
```

I would like to work on this if the contribution makes sense.

### PR DESCRIPTION
**Pull request by ishitta-iyer**
## Description

Adds sparse-label support for categorical focal crossentropy through:

- `keras.losses.sparse_categorical_focal_crossentropy`
- `keras.losses.SparseCategoricalFocalCrossentropy`

The implementation supports logits, scalar or per-class `alpha`, configurable `gamma`, `ignore_class`, arbitrary class axes, mixed precision, masking, symbolic tensors, and compiled execution across Keras backends.

It includes public API exports, serialization support, and comprehensive tests covering correctness, invalid labels, ignored classes, dtype boundaries, dynamic shapes, and custom axes.

Fixes #23442.

### Testing

- TensorFlow, JAX, NumPy, and Torch loss test suites
- TensorFlow XLA, JAX `jit`, and Torch `fullgraph`
- Public API and serialization checks
- Ruff lint and formatting checks

### AI-assisted contribution disclosure

I used an AI coding assistant during testing and review. I manually wrote, reviewed and tested the complete patch, understand its behavior, and accept responsibility for this contribution.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

**Comment by ishitta-iyer** [comment]
i just signed the CLA as well so that check should pass when you rerun the tests!

File: keras/src/losses/losses.py [hunk 1]
```diff
@@ -1174,6 +1175,149 @@ def get_config(self):
...
+            sparse_categorical_focal_crossentropy,
+            name=name,
+            reduction=reduction,
+            dtype=dtype,
+            alpha=alpha,
+            gamma=gamma,
+            from_logits=from_logits,
+            ignore_class=ignore_class,
+            axis=axis,
+        )
```
[the remark this answers is not in the document]
Comment (jeffcarp) [reply keras/src/losses/losses.py:1273]: Please fix

File: keras/src/losses/losses.py [hunk 2]
```diff
@@ -1174,6 +1175,149 @@ def get_config(self):
...
+    def call(self, y_true, y_pred):
+        # Shape handling for this loss is relative to its configurable class
+        # axis. Bypass LossFunctionWrapper's last-axis rank adjustment and let
+        # the focal loss perform its axis-aware validation and squeeze.
+        return self.fn(y_true, y_pred, **self._fn_kwargs)
+
+    def get_config(self):
+        config = super().get_config()
+        config.pop("fn")
+        return config
```
[the remark this answers is not in the document]
Comment (jeffcarp) [reply keras/src/losses/losses.py:1331]: Please fix

File: keras/src/losses/losses_test.py [hunk 3]
```diff
@@ -1993,6 +1995,424 @@ def test_dtype_arg(self):
...
+                ):
+                    losses.sparse_categorical_focal_crossentropy(
+                        y_true, y_pred, alpha=alpha
+                    )
+
+    def test_narrow_integer_labels(self):
+        for dtype, num_classes, label in (
+            ("uint8", 256, 255),
+            ("int8", 128, 127),
+        ):
```
Comment (jeffcarp) [inline keras/src/losses/losses_test.py]: Can you use a `@parameterized` test for this instead? (Here and elsewhere `self.subTest` is used)

File: keras/src/losses/losses_test.py [hunk 4]
```diff
@@ -1993,6 +1995,424 @@ def test_dtype_arg(self):
...
+        )
+        config = loss_obj.get_config()
+        self.assertEqual(config["alpha"], [0.1, 0.2, 0.7])
+        self.assertEqual(config["gamma"], 1.5)
+        self.assertEqual(config["from_logits"], True)
+        self.assertEqual(config["ignore_class"], 255)
+        self.assertEqual(config["axis"], 1)
+        self.run_class_serialization_test(loss_obj)
+
+    def test_matches_categorical_focal_crossentropy(self):
```
Comment (jeffcarp) [inline keras/src/losses/losses_test.py]: Could you also convert this and the other test methods that use manual `for` loops (`test_out_of_range_labels_produce_nan`, `test_out_of_range_ignore_class`, `test_alpha_shape_validation`, `test_class_preserves_labels_with_low_precision_dtype`) to use `@parameterized.named_parameters` from `absl.testing.parameterized`?

File: keras/src/losses/losses_test.py [hunk 5]
```diff
@@ -1993,6 +1995,424 @@ def test_dtype_arg(self):
...
+        alpha = np.array([0.2, 0.3, 0.5], dtype="float32")
+        expected = losses.sparse_categorical_crossentropy(
+            y_true, y_pred
+        ) * backend.convert_to_tensor(alpha[y_true])
+        result = losses.sparse_categorical_focal_crossentropy(
+            y_true, y_pred, alpha=alpha, gamma=0.0
+        )
+        self.assertAllClose(result, expected)
+
+    def test_ignore_class(self):
```
Comment (jeffcarp) [inline keras/src/losses/losses_test.py:2084]: Could we also add a test case here verifying that an incoming Keras mask on `y_pred` (via `backend.set_keras_mask(y_pred, ...)`) correctly combines with the `ignore_class` mask in `SparseCategoricalFocalCrossentropy.__call__`?

**Review by jeffcarp — CHANGES_REQUESTED** [review]
Thanks! Left some comments

**Comment by ishitta-iyer** [comment]
> Thanks! Left some comments

@jeffcarp made the changes!

**Review by ishitta-iyer — COMMENTED** [review]
made the changes @jeffcarp

**Review by hertschuh — APPROVED** [review]
Thanks!

File: keras/src/losses/losses_test.py [hunk 6]
```diff
@@ -1993,6 +1996,453 @@ def test_dtype_arg(self):
...
+        )
+        self.assertAllEqual(
+            backend.get_keras_mask(function_result), [True, False, True]
+        )
+        result = loss_obj(y_true, y_pred)
+        expected_valid = losses.sparse_categorical_focal_crossentropy(
+            y_true[[0, 2]], y_pred[[0, 2]]
+        )
+        self.assertAllClose(result, [expected_valid[0], 0.0, expected_valid[1]])
+
```
Comment (hertschuh) [inline keras/src/losses/losses_test.py]: This is failing on Torch. I think you can just do:
```python
self.assertAllClose(result, backend.convert_to_tensor([expected_valid[0], 0.0, expected_valid[1]]))
```

File: keras/src/losses/losses_test.py [hunk 7]
```diff
@@ -1993,6 +1996,453 @@ def test_dtype_arg(self):
...
+        )
+
+        result = losses.SparseCategoricalFocalCrossentropy(
+            ignore_class=255, reduction=None
+        )(y_true, y_pred)
+        expected = losses.sparse_categorical_focal_crossentropy(
+            np.array([2]), y_pred_values[2:]
+        )
+        self.assertAllClose(result, [0.0, 0.0, expected[0]])
+
```
Comment (hertschuh) [inline keras/src/losses/losses_test.py]: This is failing on Torch. I think you can just do:
```python
self.assertAllClose(result, backend.convert_to_tensor([0.0, 0.0, expected[0]]))
```

**Review by hertschuh — CHANGES_REQUESTED** [review]

File: keras/src/losses/losses_test.py (hunk 7 above)
Comment (ishitta-iyer) [reply keras/src/losses/losses_test.py]: done

File: keras/src/losses/losses_test.py (hunk 6 above)
Comment (ishitta-iyer) [reply keras/src/losses/losses_test.py]: done

**Comment by ishitta-iyer** [comment]
@hertschuh made the changes!

File: keras/src/losses/losses_test.py [hunk 8]
```diff
@@ -1993,6 +1996,460 @@ def test_dtype_arg(self):
...
+            ),
+        )
+
+        weighted_result = loss_obj(
+            y_true, y_pred, sample_weight=np.array([2.0, 100.0, 3.0])
+        )
+        self.assertAllClose(
+            weighted_result,
+            [expected_valid[0] * 2.0, 0.0, expected_valid[1] * 3.0],
+        )
```
Comment (jeffcarp) [inline keras/src/losses/losses_test.py:2116]: This will also fail on Torch GPU for the same reason as above (`expected_valid[0] * 2.0` is a CUDA tensor inside a Python list, which fails when converted via `np.array(...)`). Please wrap this in `backend.convert_to_tensor(...)` as well:

```suggestion
        self.assertAllClose(
            weighted_result,
            backend.convert_to_tensor(
                [expected_valid[0] * 2.0, 0.0, expected_valid[1] * 3.0]
            ),
        )
```

**Review by jeffcarp — CHANGES_REQUESTED** [review]
Thanks, one more thing:

File: keras/src/losses/losses_test.py (hunk 8 above)
Comment (ishitta-iyer) [reply keras/src/losses/losses_test.py:2117]: done

**Comment by ishitta-iyer** [comment]
> Thanks, one more thing:

Done @jeffcarp
